src/io/flutter/inspector/InspectorService.java

```diff
@@ -119,27 +129,235 @@ public FlutterApp getApp() {
+   * application is ready, the next Flutter.Frame event may never come as no
+   * new frames will be triggered to draw unless something changes in the UI.
+   */
+  public CompletableFuture<Boolean> isWidgetTreeReady() {
+    if (isDaemonApiSupported) {
+      return invokeServiceMethodDaemonNoGroup("isWidgetTreeReady", new HashMap<>())
+        .thenApplyAsync((JsonElement element) -> element.getAsBoolean() == true);
+    }
+    else {
+      return invokeServiceMethodObservatoryNoGroup("isWidgetTreeReady")
+        .thenApplyAsync((InstanceRef ref) -> "true".equals(ref.getValueAsString()));
+    }
+  }
+
+  CompletableFuture<JsonElement> invokeServiceMethodDaemonNoGroup(String methodName, List<String> args) {
+    final Map<String, Object> params = new HashMap<>();
+    for (int i = 0; i < args.size(); ++i) {
+      params.put("arg" + i, args.get(i));
+    }
+    return invokeServiceMethodDaemonNoGroup(methodName, params);
+  }
+
+
+  public CompletableFuture<Void> setPubRootDirectories(List<String> rootDirectories) {
+    // TODO(jacobr): remove call to hasServiceMethod("setPubRootDirectories") after
+    // the `setPubRootDirectories` method has been in two revs of the Flutter Alpha
+    // channel. The feature is expected to have landed in the Flutter dev
+    // chanel on March 2, 2018.
+    if (!hasServiceMethod("setPubRootDirectories")) {
+      return CompletableFuture.completedFuture(null);
+    }
+
+    if (isDaemonApiSupported) {
+      return invokeServiceMethodDaemonNoGroup("setPubRootDirectories", rootDirectories).thenApplyAsync((ignored) -> null);
+    }
+    else {
+      // TODO(jacobr): remove this call as soon as
+      // `ext.flutter.inspector.*` has been in two revs of the Flutter Beta
+      // channel. The feature landed in the Flutter dev chanel on
+      // April 16, 2018.
+      final JsonArray jsonArray = new JsonArray();
+      for (String rootDirectory : rootDirectories) {
+        jsonArray.add(rootDirectory);
+      }
+      return getInspectorLibrary().eval(
+        "WidgetInspectorService.instance.setPubRootDirectories(" + new Gson().toJson(jsonArray) + ")", null, null)
+        .thenApplyAsync((instance) -> null);
+    }
+  }
+
+  CompletableFuture<InstanceRef> invokeServiceMethodObservatoryNoGroup(String methodName) {
+    return getInspectorLibrary().eval("WidgetInspectorService.instance." + methodName + "()", null, null);
+  }
+
+  CompletableFuture<JsonElement> invokeServiceMethodDaemonNoGroup(String methodName, Map<String, Object> params) {
+    return getApp().callServiceExtension("ext.flutter.inspector." + methodName, params).thenApply((JsonObject json) -> {
+      if (json.has("errorMessage")) {
+        String message = json.get("errorMessage").getAsString();
+        throw new RuntimeException(methodName + " -- " + message);
+      }
+      return json.get("result");
     });
   }
 
+public class ObjectGroup implements Disposable {
```

Comment:
All the core functionality that was under InspectorService is now under ObjectGroup.
An ObjectGroup provides an Arena to manage lifecycles for objects referenced by the inspector. I've improperly indented this class for now to try to avoid making this code more bloated.
As soon as disposed is called on a group, all futures returned by the group will start returning null and as long as safeWhenComplete is used to await futures, you will
simply never get a callback if the group is disposed.